OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -536,29 +222,30 @@ public void SetVisualPosition(Actor self, WPos pos)
 			CenterPosition = pos;
 			self.World.UpdateMaps(self, this);
 
+			var locoInfo = Info.LocomotorInfo as SubterraneanLocomotorInfo;
+			if (locoInfo == null)
+				return;
+
 			// HACK: The submerging conditions must be applied part way through a move, and this is the only method that gets called
```

Comment:
Could we extract this out by createing an `INotifyVisualPositionChanged` that is implemented by the `GrantConditionOnSubterraneanLayer`?